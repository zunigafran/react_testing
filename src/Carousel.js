import { useState } from "react";
import "./Carousel.css";
import Card from "./Card";

function Carousel({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0);
  const total = photos.length;

  const currCard = photos[currCardIdx];

  // Function to increment currCardIdx
  function goForward() {
    setCurrCardIdx((currCardIdx + 1) % total);
  }

  // Function to decrement currCardIdx
  function goBackward() {
    setCurrCardIdx((currCardIdx - 1 + total) % total);
  }

  const isAtFirstImage = currCardIdx === 0;
  const isAtLastImage = currCardIdx === total - 1;

  return (
    <div className="Carousel">
      <h1>{title}</h1>
      <div className="Carousel-main">
        <i
          className="bi bi-arrow-left-circle"
          onClick={goBackward}
          style={{ visibility: isAtFirstImage ? "hidden" : "visible" }}
        />
        <Card
          caption={currCard.caption}
          src={currCard.src}
          currNum={currCardIdx + 1}
          totalNum={total}
        />
        <i
          className="bi bi-arrow-right-circle"
          onClick={goForward}
          style={{ visibility: isAtLastImage ? "hidden" : "visible" }}
        />
      </div>
    </div>
  );
}

export default Carousel;
