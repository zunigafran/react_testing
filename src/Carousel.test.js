import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";
import TEST_IMAGES from "./_testCommon.js";

describe("Carousel Component", () => {
  it("should render without errors", () => {
    const { container } = render(
      <Carousel photos={TEST_IMAGES} title="images for testing" />
    );
    expect(container).toMatchSnapshot();
  });

  it("works when you click on the right arrow", () => {
    const { container } = render(
      <Carousel photos={TEST_IMAGES} title="images for testing" />
    );
    // expect the first image to show, but not the second
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).not.toBeInTheDocument();

    // move forward in the carousel
    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);

    // expect the second image to show, but not the first
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).toBeInTheDocument();
  });

  it("left arrow moves to the previous image (BUGGY)", () => {
    const { container } = render(
      <Carousel photos={TEST_IMAGES} title="images for testing" />
    );

    // Click right arrow to go to the second image
    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);

    // expect the second image to show, but not the first
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).toBeInTheDocument();

    // Click left arrow to go back to the first image (BUGGY - Expected behavior is to go back to the first image)
    const leftArrow = container.querySelector(".bi-arrow-left-circle");
    fireEvent.click(leftArrow);

    // This assertion is expected to fail due to the bug
    // When the bug is fixed, the following expectations should pass
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).not.toBeInTheDocument();
  });
});