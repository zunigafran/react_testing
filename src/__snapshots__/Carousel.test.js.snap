// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Component should render without errors 1`] = `
<div>
  <div
    class="Carousel"
  >
    <h1>
      images for testing
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="bi bi-arrow-left-circle"
        style="visibility: hidden;"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          testing image 1
        </h4>
        <img
          alt="testing image 1"
          class="Card-image"
          src="test1.com"
        />
        <small
          class="Card-small"
        >
          Image 
          1
           of 
          3
          .
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
        style="visibility: visible;"
      />
    </div>
  </div>
</div>
`;
